# Introduction to the OOI M2M Interface (aka the API)
*Written by Sage Lichtenwalner, Rutgers University, May 21, 2018*

*Revised by Lori Garzio (to work with Pangeo), Rutgers University, July 10, 2018*

In this example, we will demonstrate how to access and use the OOI M2M interface to retrieve information about the instruments in the OOI database as well as how to retrieve data from those instruments.

For most of this example, I'm going to use my favorite instrument, the [30m CTD on Global Papa Flanking Mooring B](http://ooi.visualocean.net/instruments/view/GP03FLMB-RIM01-02-CTDMOG060).

# Getting Started with the API

In order to use the API, you will need to have an OOI Net account, that is, an account on the [OOI Data Portal](https://ooinet.oceanobservatories.org).  Once you have an account, you will need your API Username and Token.  To get your API information, follow the steps below, or check out this [video](https://www.youtube.com/watch?time_continue=150&v=9YSIrDu4l24).

* If needed, create a new account on ooinet.oceanobservatories.org, using either your email address or use the CILogon button with your academic or Google account.
* Log in
* Click on your email address on the right side of the navigation bar to open up the pull-down menu.
* Click on the "User Profile" in the drop down.
* Copy and save the following data from your user profile: API Username and API Token. The API Username is similar to “OOIAPI-QTULEV4STGAS35”. The API Token is similar to “YYP2X2W3SOW”.

For this example, we will include the API login information as variables so we can use them later.


In [1]:
USERNAME = 'YOUR API USERNAME'
TOKEN =  'YOUR API TOKEN'

We will also add a few Python libraries that we will need in our environment

In [2]:
# First, we need to add some more Python libraries
import requests
import datetime

To access the OOI API, we will use the [python requests library](http://docs.python-requests.org/en/master/), which is very easy to use.  

> `r = requests.get(url, params=parameters, auth=('user', 'pass'))`

All we have to do is specify the URL we want to access, along with our login information.  In some cases, we will also include an additional set of optional parameters using the "params" option.  These "GET" parameters are typically found at the end of a URL following the format:
> `http://example.com/page.htm?var1=a&var2=b`

Thanks to the requests library, we don't have to worry about getting all the question marks and ampersands in the right places, we can just specify an array of options.


To see how this really works in practice, let's dive into some examples.


## Sensor Information

Perhaps the most useful API is the **Sensor Inventory**.  This endpoint essentially works like a tree, allowing you to navigate through the OOI instrument hierarchy.  Here is the starting point.

In [3]:
SENSOR_BASE_URL = 'https://ooinet.oceanobservatories.org/api/m2m/12576/sensor/inv/'

Let's go ahead and take a look at what we get back from this URL.  Thanks to the requests library, we have a number of different ways to deal with the output.

In [4]:
r = requests.get(SENSOR_BASE_URL, auth=(USERNAME, TOKEN))
print( r.status_code )
print( r.headers )
print( r.encoding )
print( r.text )
print( r.json() )

200
{'Server': 'nginx', 'Date': 'Tue, 10 Jul 2018 14:04:44 GMT', 'Content-Type': 'text/html; charset=utf-8', 'Transfer-Encoding': 'chunked', 'Strict-Transport-Security': 'max-age=63072000; includeSubdomains; ', 'X-Frame-Options': 'SAMEORIGIN', 'X-XSS-Protection': '1; mode=block', 'Content-Encoding': 'gzip'}
utf-8
[ "CE01ISSM", "CE01ISSP", "CE02SHBP", "CE02SHSM", "CE02SHSP", "CE04OSBP", "CE04OSPD", "CE04OSPI", "CE04OSPS", "CE04OSSM", "CE05MOAS", "CE06ISSM", "CE06ISSP", "CE07SHSM", "CE07SHSP", "CE09OSPM", "CE09OSSM", "CP01CNPM", "CP01CNSM", "CP01CNSP", "CP02PMCI", "CP02PMCO", "CP02PMUI", "CP02PMUO", "CP03ISPM", "CP03ISSM", "CP03ISSP", "CP04OSPM", "CP04OSSM", "CP05MOAS", "GA01SUMO", "GA02HYPM", "GA03FLMA", "GA03FLMB", "GA05MOAS", "GI01SUMO", "GI02HYPM", "GI03FLMA", "GI03FLMB", "GI05MOAS", "GP02HYPM", "GP03FLMA", "GP03FLMB", "GP05MOAS", "GS01SUMO", "GS02HYPM", "GS03FLMA", "GS03FLMB", "GS05MOAS", "RS01OSBP", "RS01SBPS", "RS01SHBP", "RS01SHDR", "RS01SLBS", "RS01SUM1", "RS01SUM2", "RS03ASHS",

As we can see, the returned status code is 200, which means our request was OK and and information was returned.  

The headers and encoding tell us a little bit more about the format of the information sent back.  Because this is a simple text response, there's nothing profound here, but you might find these handy in working with other APIs.

We can retrieve the actual data from the response using **r.text**. But because this dataset is actually in JSON format, we can use the **r.json()** method to automatically turn the returned data into a structured object we can use in Python.  

For example, let's save the output as a variable and then loop through it an print out each entry.

In [5]:
sites = r.json()
for site in sites:
  print( site )

CE01ISSM
CE01ISSP
CE02SHBP
CE02SHSM
CE02SHSP
CE04OSBP
CE04OSPD
CE04OSPI
CE04OSPS
CE04OSSM
CE05MOAS
CE06ISSM
CE06ISSP
CE07SHSM
CE07SHSP
CE09OSPM
CE09OSSM
CP01CNPM
CP01CNSM
CP01CNSP
CP02PMCI
CP02PMCO
CP02PMUI
CP02PMUO
CP03ISPM
CP03ISSM
CP03ISSP
CP04OSPM
CP04OSSM
CP05MOAS
GA01SUMO
GA02HYPM
GA03FLMA
GA03FLMB
GA05MOAS
GI01SUMO
GI02HYPM
GI03FLMA
GI03FLMB
GI05MOAS
GP02HYPM
GP03FLMA
GP03FLMB
GP05MOAS
GS01SUMO
GS02HYPM
GS03FLMA
GS03FLMB
GS05MOAS
RS01OSBP
RS01SBPS
RS01SHBP
RS01SHDR
RS01SLBS
RS01SUM1
RS01SUM2
RS03ASHS
RS03AXBS
RS03AXPD
RS03AXPS
RS03AXSM
RS03CCAL
RS03ECAL
RS03INT1
RS03INT2
SSRSPACC


Now this may seem like gibberish, but it turns out that this is a list of all of the **Sites** that are in the OOI system.  A [similar list](http://oceanobservatories.org/site-list/) can be found on the OOI website.

From this point on, we can simply add one of the returned items to the previous URL in order to drill down into the sensor inventory.  For this example, let's investigate [**GP03FLMB-RIM01-02-CTDMOG060**](http://ooi.visualocean.net/instruments/view/GP03FLMB-RIM01-02-CTDMOG060).

To make this easy, let's setup a quick function to make the request and reformat the results.

In [6]:
# A quick function to make an API request and print the results
def get_and_print_api(url):
  r = requests.get(url, auth=(USERNAME, TOKEN))
  data = r.json()
  for d in data:
    print( d )

### List of Nodes for a Site

In [7]:
get_and_print_api(SENSOR_BASE_URL+'/GP03FLMB')

RIM01
RIS01


### List of Sensors (Instruments) for a Node

In [8]:
get_and_print_api(SENSOR_BASE_URL+'/GP03FLMB/RIM01')

00-SIOENG000
02-ADCPSL007
02-CTDMOG060
02-CTDMOG061
02-CTDMOG062
02-CTDMOG063
02-CTDMOG064
02-CTDMOG065
02-CTDMOG066
02-CTDMOG067
02-CTDMOG068
02-CTDMOH069
02-CTDMOH070
02-CTDMOH071


### List of Methods for a Sensor

In [9]:
get_and_print_api(SENSOR_BASE_URL+'/GP03FLMB/RIM01/02-CTDMOG060')

bad_recovered_host
bad_telemetered
recovered_host
recovered_inst
telemetered


### List of Data Streams for a Method

In [10]:
get_and_print_api(SENSOR_BASE_URL+'/GP03FLMB/RIM01/02-CTDMOG060/telemetered')

ctdmo_ghqr_sio_mule_instrument
ctdmo_ghqr_sio_offset


At this point, we have all the information we need to make a data request.  All we have to do is go down one additional level by adding the stream name to the request we made above.  **However, don't be tempted to do this.**  If you do, you will accidentally make an "asynchronous" request (aka a request for downloadable netcdf files) for all of the data in the system for this data stream (unless, of course, this is what you want to do).

Instead, of doing that, let's dive into how to make more precise data requests.

# Synchronous Data Requests 
* Results returned as JSON

To make a synchronous data request, we basically construct a URL using all the pieces we discovered above, like the following:
> /sensor/inv/{subsite}/{node}/{sensor}/{method}/{stream}

In order to make the code clear, we're going to setup several variables and then use the **join()** function to concatenate all of the variables together with slashes.

We are also going to use a **params** array to specify a start and ending date/time for our request, as well as a limit parameter to minimize the amount of data points we get back.

Keep in mind that synchronous requests are **limited to 20,000 data points** within the time range you specify.  If the dataset you are requesting has more data points available, the data will be decimated (that is, roughly downsampled in a semi-random way).  To request all of the available data, you should make an asynchronous request (see below).  Alternatively, you could make a sequence of synchronous requests for a shorter time range, and then aggregate the results.  For a real-time plotting script, this actually makes more sense (but that's another example).

In [11]:
# Instrument Information
site = 'GP03FLMB'
node = 'RIM01'
instrument = '02-CTDMOG060'
method = 'telemetered'
stream = 'ctdmo_ghqr_sio_mule_instrument'

data_request_url ='/'.join((SENSOR_BASE_URL,site,node,instrument,method,stream))

params = {
  'beginDT':'2016-10-01T00:00:00.000Z',
  'endDT':'2016-11-01T00:00:00.000Z',
  'limit':1000,   
}

Now that everything is setup, let's make the request. Here we will return the length of the returned data array.

In [17]:
r = requests.get(data_request_url, params=params, auth=(USERNAME, TOKEN))
data = r.json()
print( len(data) )

6


Let's take a look at the first item in the list.  From this, we can see that each item is a dictionary of every variable this instrument measures.

In [13]:
data[0]

{'practical_salinity': 32.395352318774655,
 'ctd_time': 529228801,
 'density_qc_executed': 29,
 'driver_timestamp': 3722685717.0375433,
 'conductivity': 397190,
 'ctdmo_seawater_pressure_qc_executed': 29,
 'practical_salinity_qc_results': 29,
 'temperature': 190588,
 'ctdmo_seawater_conductivity_qc_results': 29,
 'density': 1025.243494104906,
 'ctdmo_seawater_conductivity_qc_executed': 29,
 'ctdmo_seawater_pressure_qc_results': 29,
 'ctdmo_seawater_temperature': 9.058800000000002,
 'internal_timestamp': 3684902401.0,
 'pk': {'node': 'RIM01',
  'stream': 'ctdmo_ghqr_sio_mule_instrument',
  'subsite': 'GP03FLMB',
  'deployment': 4,
  'time': 3684902401.0,
  'sensor': '02-CTDMOG060',
  'method': 'telemetered'},
 'ctdmo_seawater_conductivity': 3.4719,
 'ctdmo_seawater_temperature_qc_executed': 29,
 'ingestion_timestamp': 3722685723.77,
 'port_timestamp': 0.0,
 'pressure': 4873,
 'ctdmo_seawater_temperature_qc_results': 29,
 'ctdmo_seawater_pressure': 37.088139230566796,
 'preferred_timesta

### What it looks like when there isn't any data?
Now, if we specified an time range that didn't have any data, this is what we would get.

In [14]:
params = {
  'beginDT':'2017-10-24T00:00:00.000Z',
  'endDT':'2017-10-25T00:00:00.000Z',
  'limit':1000,   
}

r = requests.get(data_request_url, params=params, auth=(USERNAME, TOKEN))
data = r.json()
print( len(data) )
data

2


{'message': {'status': '{\n  "message": "Query returned no results for primary stream", \n  "requestUUID": "e14a893a-0430-4440-9224-76a62e12c072"\n}\n'},
 'status_code': 404}

Note that in this case, the returned data array is only 2 items long, and the returned status_code is **404**, instead of **200**.  If you need to verify whether data was returned, you can use this info to add error checking to your script.

### How do you limit the number of parameters returned?
To limit the amount of data in the returned JSON array, we can specify just the parameters we want.  However, to do this, we need to figure out the parameter IDs (see below).

In [15]:
params = {
  'beginDT':'2016-10-01T00:00:00.000Z',
  'endDT':'2016-11-01T00:00:00.000Z',
  'limit':1000,
  'parameters':'7,13' #time=7, salinity=13, temperature=2927, pressure=2926
}

r = requests.get(data_request_url, params=params, auth=(USERNAME, TOKEN))
data = r.json()
print( len(data) )
data[1]

20


{'practical_salinity': 32.57040101955179,
 'practical_salinity_qc_results': 29,
 'pk': {'node': 'RIM01',
  'stream': 'ctdmo_ghqr_sio_mule_instrument',
  'subsite': 'GP03FLMB',
  'deployment': 4,
  'time': 3684916801.0,
  'sensor': '02-CTDMOG060',
  'method': 'telemetered'},
 'practical_salinity_qc_executed': 29,
 'time': 3684916801.0}

# Asynchnorous Data Requests 
* Results returned as downloadable NetCDF files

Synchronous requests are great when you just want to grab a little bit of data quickly. Thus, it is very useful for applications like interactive graphs on a web site.

However, if you want a larger and more complete dataset, you will need to create an Asynchronous request, which take longer to process, sometimes up to an hour or more depending on how much data you are requesting and what else the system is doing.  The end result will be a web accessible directory of NetCDF data files.

This is essentially the same functionality as requesting a download from the OOI Data Portal, but with the API you can easily create one or more requests in an automated way.

The request format is essentially the same as above, except we can drop the **limit** option from the params array.  We could also omit the start and end dates, if we want to grab all available data.

By default, asynchronous requests will return NetCDF files, but you could also specify csv or json, using the **format** option.  Optionally, you can also specify **include_provenance** and **include_annotations** which will include separate text files in the output directory with that information.

In [16]:
# Instrument Information
site = 'GP03FLMB'
node = 'RIM01'
instrument = '02-CTDMOG060'
method = 'telemetered'
stream = 'ctdmo_ghqr_sio_mule_instrument'

# Create the request URL
data_request_url ='/'.join((SENSOR_BASE_URL,site,node,instrument,method,stream))

# All of the following are optional
params = {
  'beginDT':'2016-10-01T00:00:00.000Z',
  'endDT':'2016-11-01T00:00:00.000Z',
  'format':'application/netcdf',
  'include_provenance':'true',
  'include_annotations':'true'
}

Now let's send the request.  (Note, these lines are commented out, to prevent accidental resubmission when running through the entire notebook quickly.)

In [18]:
# r = requests.get(data_request_url, params=params, auth=(USERNAME, TOKEN))
# data = r.json()

While the Synchronous request returned an array of actual data, this time we will just get some URLs and some other metadata about our request.

In [19]:
data

{'requestUUID': 'dd91c0cd-ecb8-4334-b4cc-7d5197120278',
 'outputURL': 'https://opendap.oceanobservatories.org/thredds/catalog/ooi/ooidatateam@gmail.com/20180710T141111-GP03FLMB-RIM01-02-CTDMOG060-telemetered-ctdmo_ghqr_sio_mule_instrument/catalog.html',
 'allURLs': ['https://opendap.oceanobservatories.org/thredds/catalog/ooi/ooidatateam@gmail.com/20180710T141111-GP03FLMB-RIM01-02-CTDMOG060-telemetered-ctdmo_ghqr_sio_mule_instrument/catalog.html',
  'https://opendap.oceanobservatories.org/async_results/ooidatateam@gmail.com/20180710T141111-GP03FLMB-RIM01-02-CTDMOG060-telemetered-ctdmo_ghqr_sio_mule_instrument'],
 'sizeCalculation': 4080,
 'timeCalculation': 60,
 'numberOfSubJobs': 2}

The first URL in the **allURLs** key points to the THREDDS server, which allows for programmatic data access without downloading the entire file.

In [20]:
print(data['allURLs'][0])

https://opendap.oceanobservatories.org/thredds/catalog/ooi/ooidatateam@gmail.com/20180710T141111-GP03FLMB-RIM01-02-CTDMOG060-telemetered-ctdmo_ghqr_sio_mule_instrument/catalog.html


The second URL in the **allURLs** key provides a direct link to a web server which you can use to quickly download files if you don't want to go through THREDDS.

In [21]:
print(data['allURLs'][1])

https://opendap.oceanobservatories.org/async_results/ooidatateam@gmail.com/20180710T141111-GP03FLMB-RIM01-02-CTDMOG060-telemetered-ctdmo_ghqr_sio_mule_instrument


### How can you check when a request is complete?

We can use the second URL to check if a status.txt file has been written to the location. If true, then the request has completed and all data have been delivered to the THREDDS server.

The following for loop will poll the location for the status.txt files 1000 times, once every 0.5 seconds. Therefore, it will quit after about 8.3 minutes. If you are requesting a very large and dense dataset (3 years of BOTPT data collected at 20 Hz, for example), it may take a bit longer to complete the request, so you will want to bump up the retry range or lower the sleep interval.

In [25]:
import time

In [26]:
%%time
check_complete = data['allURLs'][1] + '/status.txt'
for i in range(1000): 
    r = requests.get(check_complete)
    if r.status_code == requests.codes.ok:
        print('request completed')
        break
    else:
        time.sleep(.5)

request completed
CPU times: user 15 ms, sys: 1 ms, total: 16 ms
Wall time: 150 ms


# Sensor Metadata

Getting data is nice, but one of the other big advantages to using the API is that you can query the system about all kinds of metadata.  To start, let's see what sensor metadata is available.

This query looks a lot like the above data requests, except we're explicitly asking for just the metadata.

The following metadata requests are possible:

```
/sensor/inv/{subsite}/{node}/{sensor}/metadata
/sensor/inv/CE01ISSM/MFC31/00-CPMENG000/metadata/times
/sensor/inv/CE01ISSM/MFC31/00-CPMENG000/metadata/parameters

```
The first request grabs both times and parameters, the others specify just a single request.


In [27]:
# Instrument Information
site = 'GP03FLMB'
node = 'RIM01'
instrument = '02-CTDMOG060'
method = 'telemetered'
stream = 'ctdmo_ghqr_sio_mule_instrument'

# Create the request URL
data_request_url ='/'.join((SENSOR_BASE_URL,site,node,instrument,'metadata'))

In [28]:
r = requests.get(data_request_url, params=params, auth=(USERNAME, TOKEN))
data = r.json()

In [29]:
# Print out the response using PrettyPrint
# import pprint
# pp = pprint.PrettyPrinter(indent=2)
# pp.pprint(data)

import json

# Print out the response using json.dumps()
# print json.dumps(data, indent=4);

# Print out just one element
print( json.dumps(data['parameters'][0], indent=4) );
print( json.dumps(data['times'][0], indent=4) );

{
    "type": "DOUBLE",
    "stream": "ctdmo_ghqr_instrument_recovered",
    "units": "seconds since 1900-01-01",
    "fillValue": "-9999999",
    "pdId": "PD7",
    "particleKey": "time",
    "unsigned": false,
    "shape": "SCALAR"
}
{
    "stream": "ctdmo_ghqr_offset_recovered",
    "method": "recovered_host",
    "count": 1092,
    "endTime": "2017-07-16T20:07:20.000Z",
    "beginTime": "2013-07-25T22:07:18.000Z"
}


The response includes a list of every parameter for every data stream produced by the instrument.  There may be several entries with the same parameter, if multiple streams produce the same one.

The times array, includes the start and end times of the entire data record for each stream.  There may still be gaps, but this is useful for knowing how up to date a data stream is (i.e. is it close to real time, or how recent was the last recovered data available?)

Let's put all of the parameters into a single table, and then filter the results to just the telemetered stream we were interested in above.

In [30]:
import pandas as pd

# Convert the parameters array into a DataFrame
parameters = pd.DataFrame(data['parameters'],
                          columns=['particleKey','pdId','units','stream'])

# Filter out the parameters for a single stream
parameters = parameters[parameters['stream']=='ctdmo_ghqr_sio_mule_instrument']

# Print out selected columns
print( parameters[['particleKey','pdId','units']] )


                    particleKey    pdId                     units
27                         time     PD7  seconds since 1900-01-01
28               port_timestamp    PD10  seconds since 1900-01-01
29             driver_timestamp    PD11  seconds since 1900-01-01
30           internal_timestamp    PD12  seconds since 1900-01-01
31          preferred_timestamp    PD16                         1
32                  temperature   PD193                    counts
33                 conductivity   PD194                    counts
34                     pressure   PD195                    counts
35                     ctd_time   PD198  seconds since 2000-01-01
36          ingestion_timestamp   PD863  seconds since 1900-01-01
37                 inductive_id  PD1558                         1
38     sio_controller_timestamp  PD2174  seconds since 1970-01-01
39      ctdmo_seawater_pressure  PD2926                      dbar
40   ctdmo_seawater_temperature  PD2927                        ºC
41  ctdmo_

# Parameter & Stream Information

We could have gotten a similar list by using a different API endpoint, to retrieve information on a specific stream.

In [31]:
# Retrieve information for a Preload Stream given its name. 
url = 'https://ooinet.oceanobservatories.org/api/m2m/12575/stream/byname/ctdmo_ghqr_sio_mule_instrument'
r = requests.get(url, auth=(USERNAME, TOKEN))
data = r.json()

# Convert to Data Frame
parameters = pd.DataFrame(data['parameters'])
parameters[['id','name','data_product_type','display_name','unit']]

id                         name              data_product_type  \
0      7                         time                           None   
1     10               port_timestamp                           None   
2     11             driver_timestamp                           None   
3     12           internal_timestamp                           None   
4     16          preferred_timestamp                           None   
5    193                  temperature  {'value': 'Unprocessed Data'}   
6    194                 conductivity  {'value': 'Unprocessed Data'}   
7    195                     pressure  {'value': 'Unprocessed Data'}   
8    198                     ctd_time                           None   
9    863          ingestion_timestamp                           None   
10  1558                 inductive_id                           None   
11  2174     sio_controller_timestamp                           None   
12  2926      ctdmo_seawater_pressure      {'value': 'Science Data'}   
13  2927   ctdmo_seawater_temperature      {'value': 'Science Data'}   
14  2928  ctdmo_seawater_conductivity      {'value': 'Science Data'}   
15     5                      density      {'value': 'Science Data'}   
16    13           practical_salinity      {'value': 'Science Data'}   

                         display_name                                   unit  
0                           Time, UTC  {'value': 'seconds since 1900-01-01'}  
1                 Port Timestamp, UTC  {'value': 'seconds since 1900-01-01'}  
2               Driver Timestamp, UTC  {'value': 'seconds since 1900-01-01'}  
3             Internal Timestamp, UTC  {'value': 'seconds since 1900-01-01'}  
4                 Preferred Timestamp                         {'value': '1'}  
5    Seawater Temperature Measurement                    {'value': 'counts'}  
6   Seawater Conductivity Measurement                    {'value': 'counts'}  
7       Seawater Pressure Measurement                    {'value': 'counts'}  
8                           Time, UTC  {'value': 'seconds since 2000-01-01'}  
9            Ingestion Timestamp, UTC  {'value': 'seconds since 1900-01-01'}  
10                       Inductive ID                         {'value': '1'}  
11                SIO Controller Time  {'value': 'seconds since 1970-01-01'}  
12                  Seawater Pressure                      {'value': 'dbar'}  
13               Seawater Temperature                        {'value': 'ºC'}  
14              Seawater Conductivity                     {'value': 'S m-1'}  
15                   Seawater Density                    {'value': 'kg m-3'}  
16                 Practical Salinity                         {'value': '1'}

And to get information on a specific parameter...

In [32]:
# Retrieve information for a Preload Parameter given its identifier.
url = 'https://ooinet.oceanobservatories.org/api/m2m/12575/parameter/13'
r = requests.get(url, auth=(USERNAME, TOKEN))
data = r.json()

print( json.dumps(data, indent=4) )

{
    "name": "practical_salinity",
    "display_name": "Practical Salinity",
    "standard_name": "sea_water_practical_salinity",
    "description": "Salinity is generally defined as the concentration of dissolved salt in a parcel of seawater. Practical Salinity is a more specific unitless quantity calculated from the conductivity of seawater and adjusted for temperature and pressure. It is approximately equivalent to Absolute Salinity (the mass fraction of dissolved salt in seawater) but they are not interchangeable.",
    "id": 13,
    "data_product_identifier": "PRACSAL_L2",
    "precision": 4,
    "fill_value": {
        "value": "-9999999"
    },
    "unit": {
        "value": "1"
    },
    "data_level": 2,
    "code_set": null,
    "value_encoding": {
        "value": "float32"
    },
    "parameter_type": {
        "value": "function"
    },
    "parameter_function": {
        "name": "ctd_pracsal",
        "id": 26,
        "owner": "ion_functions.data.ctd_functions",
       

# Deployments

To find out how many times an instrument has been deployed (and for which there is data in the system), we can use the Deployments API endpoint.

In [33]:
# Retrieve vocabulary information for a given instrument
DEPLOYMENT_BASE_URL = 'https://ooinet.oceanobservatories.org/api/m2m/12587/events/deployment/inv'
request_url = DEPLOYMENT_BASE_URL+'/GP03FLMB/RIM01/02-CTDMOG060'
r = requests.get(request_url, auth=(USERNAME, TOKEN))
data = r.json()

data

[1, 2, 3, 4, 5]

We can then add a specific deployment number, to retrieve detailed information on that deployment.  Alternatively, you can specify "-1" to get information about all deployments.

In [36]:
# Retrieve vocabulary information for a given instrument
request_url = DEPLOYMENT_BASE_URL+'/GP03FLMB/RIM01/02-CTDMOG060/5'
r = requests.get(request_url, auth=(USERNAME, TOKEN))
data = r.json()
# print( json.dumps(data, indent=4) )

Note that due to a quirk in the API, it typically returns all calibration values for an instrument, not all of which are relevant for the specific deployment you may be interested in. This is a known issue and there is a ticket for the software developers to work on a fix.

Let's pull out the unique sensor IDs for this deployment.

In [38]:
print( data[0]['sensor']['uid'] )
print( data[0]['sensor']['assetId'] )

CGINS-CTDMOG-13422
2840


# Asset Information

OOI Reference Designators denote a specific measurement location in the OOI, for example the 30m CTD at the Papa Flanking Mooring B location.  But over time, measurements made at this "instrument location" are actually made by multiple instruments.  Each of those would have a specific Make, Model, and unique Serial Number.  We call these "assets" in the OOI system.

You can use the Asset API endpoint to query information about a specific asset, once you kow either it's ID or UID.  In the above example, we extracted both to find out the unique sensor used during a specific deployment.

In [39]:
# Request Asset Information by UID
ASSET_BASE_URL = 'https://ooinet.oceanobservatories.org/api/m2m/12587/asset'
request_url = ASSET_BASE_URL+'?uid=CGINS-CTDMOG-13422'
r = requests.get(request_url, auth=(USERNAME, TOKEN))
data = r.json()
# print( json.dumps(data, indent=4) )

This returns essentially the same information that was embedded in the deployment information request above, including sensor calibration information.

We can also find out where this particular asset was deployed over time.

In [40]:
# Get deployment digests by the unique asset identifier (UID).
ASSET_BASE_URL = 'https://ooinet.oceanobservatories.org/api/m2m/12587/asset'
request_url = ASSET_BASE_URL+'/deployments/CGINS-CTDMOG-13422'
r = requests.get(request_url, auth=(USERNAME, TOKEN))
data = r.json()

# print( json.dumps(data, indent=4) )
pd.DataFrame(data)

deployCruiseIdentifier  deploymentNumber  depth    editPhase       endTime  \
0               AT-30-01                 2   90.0  OPERATIONAL  1.468714e+12   
1                SR17-10                 5    NaN  OPERATIONAL           NaN   

   eventId  latitude   longitude         mooring_uid   node node_uid  \
0     2296  59.77067  -39.880100  CGMGI-03FLMA-00002  RIM01     None   
1     3627  50.37768 -144.514867  CGMGP-03FLMB-00005  RIM01     None   

  orbitRadius recoverCruiseIdentifier        sensor          sensor_uid  \
0        None                AR-07-01  02-CTDMOG043  CGINS-CTDMOG-13422   
1        None                    None  02-CTDMOG060  CGINS-CTDMOG-13422   

       startTime   subsite  versionNumber  waterDepth  
0  1439941680000  GI03FLMA              1      2689.0  
1  1500072540000  GP03FLMB              1      4169.0

In this case, we see that this particular instrument has been deployed twice.  Once at Global Irminger (deployment 2 of that site) and a second time at Global Papa (deployment 5).

# Calibration Information

As we've seen above, a number of API endpoints provide calibration information, including the Sensor, Deployments and Asset ones.  We could also use the following Asset endpoint to search for sensor calibrations for a particular reference designator.

In [41]:
# Simple time conversion function
def convert_time(ms):
  if ms != None:
    return datetime.datetime.utcfromtimestamp(ms/1000)
  else:
    return None

In [42]:
# Get deployment digests by the unique asset identifier (UID).
ASSET_BASE_URL = 'https://ooinet.oceanobservatories.org/api/m2m/12587/asset'
request_url = ASSET_BASE_URL+'/cal?refdes=GP03FLMB-RIM01-02-CTDMOG060'

r = requests.get(request_url, auth=(USERNAME, TOKEN))
data = r.json()

# print( json.dumps(data, indent=4) )
# data[4]
pd.DataFrame(data)

@class assetUid                        dataSource  \
0  .XDeployment     None  Load from [GP03FLMB_Deploy.xlsx]   
1  .XDeployment     None  Load from [GP03FLMB_Deploy.xlsx]   
2  .XDeployment     None  Load from [GP03FLMB_Deploy.xlsx]   
3  .XDeployment     None  Load from [GP03FLMB_Deploy.xlsx]   
4  .XDeployment     None  Load from [GP03FLMB_Deploy.xlsx]   

                                    deployCruiseInfo deployedBy  \
0  {'@class': '.CruiseInfo', 'uniqueCruiseIdentif...       None   
1  {'@class': '.CruiseInfo', 'uniqueCruiseIdentif...       None   
2  {'@class': '.CruiseInfo', 'uniqueCruiseIdentif...       None   
3  {'@class': '.CruiseInfo', 'uniqueCruiseIdentif...       None   
4  {'@class': '.CruiseInfo', 'uniqueCruiseIdentif...       None   

   deploymentNumber    editPhase  eventId                    eventName  \
0                 1  OPERATIONAL     3555  GP03FLMB-RIM01-02-CTDMOG060   
1                 2  OPERATIONAL     3573  GP03FLMB-RIM01-02-CTDMOG060   
2                 3  OPERATIONAL     3591  GP03FLMB-RIM01-02-CTDMOG060   
3                 4  OPERATIONAL     3609  GP03FLMB-RIM01-02-CTDMOG060   
4                 5  OPERATIONAL     3627  GP03FLMB-RIM01-02-CTDMOG060   

   eventStartTime    ...      \
0   1374647520000    ...       
1   1403229600000    ...       
2   1433798700000    ...       
3   1467673620000    ...       
4   1500072540000    ...       

                                             mooring  node notes  \
0  {'@class': '.XMooring', 'remoteResources': [],...  None  None   
1  {'@class': '.XMooring', 'remoteResources': [],...  None  None   
2  {'@class': '.XMooring', 'remoteResources': [],...  None  None   
3  {'@class': '.XMooring', 'remoteResources': [],...  None  None   
4  {'@class': '.XMooring', 'remoteResources': [],...  None  None   

                                   recoverCruiseInfo  recoveredBy  \
0  {'@class': '.CruiseInfo', 'uniqueCruiseIdentif...         None   
1  {'@class': '.CruiseInfo', 'uniqueCruiseIdentif...         None   
2  {'@class': '.CruiseInfo', 'uniqueCruiseIdentif...         None   
3  {'@class': '.CruiseInfo', 'uniqueCruiseIdentif...         None   
4                                               None         None   

           referenceDesignator  \
0  GP03FLMB-RIM01-02-CTDMOG060   
1  GP03FLMB-RIM01-02-CTDMOG060   
2  GP03FLMB-RIM01-02-CTDMOG060   
3  GP03FLMB-RIM01-02-CTDMOG060   
4  GP03FLMB-RIM01-02-CTDMOG060   

                                              sensor    tense versionNumber  \
0  {'@class': '.XInstrument', 'calibration': [{'@...  UNKNOWN             1   
1  {'@class': '.XInstrument', 'calibration': [{'@...  UNKNOWN             1   
2  {'@class': '.XInstrument', 'calibration': [{'@...  UNKNOWN             1   
3  {'@class': '.XInstrument', 'calibration': [{'@...  UNKNOWN             1   
4  {'@class': '.XInstrument', 'calibration': [{'@...  UNKNOWN             1   

  waterDepth  
0     4145.0  
1     4145.0  
2     4145.0  
3     4146.0  
4     4169.0  

[5 rows x 26 columns]

In [43]:
# Reformat the data into a pretty table
df = pd.DataFrame() # Setup empty array
for d in data:
  for dd in d['sensor']['calibration']:
    for ddd in dd['calData']:
      df = df.append({
        'value': ddd['value'],
        'start': convert_time(ddd['eventStartTime']),
        'stop': convert_time(ddd['eventStopTime']),
        'name': ddd['eventName'],
        'assetUid': ddd['assetUid'],
        }, ignore_index=True)
    
df = df.sort_values(by=['name'])
df

assetUid        name      start  stop         value
0    CGINS-CTDMOG-10255       CC_a0 2012-12-11  None -1.546605e-04
193  CGINS-CTDMOG-11638       CC_a0 2016-07-04  None -1.041285e-04
192  CGINS-CTDMOG-11638       CC_a0 2014-06-20  None -7.357588e-05
72   CGINS-CTDMOG-11646       CC_a0 2017-11-21  None -8.245547e-05
73   CGINS-CTDMOG-11646       CC_a0 2014-06-20  None -1.166783e-04
74   CGINS-CTDMOG-11646       CC_a0 2016-07-11  None -8.666183e-05
264  CGINS-CTDMOG-13422       CC_a0 2016-10-15  None -1.724387e-04
265  CGINS-CTDMOG-13422       CC_a0 2015-08-18  None -1.337706e-04
194  CGINS-CTDMOG-11638       CC_a0 2017-09-01  None -8.046998e-05
2    CGINS-CTDMOG-10255       CC_a0 2015-11-10  None -1.314068e-04
145  CGINS-CTDMOG-12638       CC_a0 2016-09-21  None -8.597695e-05
144  CGINS-CTDMOG-12638       CC_a0 2015-06-08  None -1.153096e-04
1    CGINS-CTDMOG-10255       CC_a0 2017-09-06  None -1.158514e-04
28   CGINS-CTDMOG-10255       CC_a1 2015-11-10  None  3.106929e-04
100  CGINS-CTDMOG-11646       CC_a1 2016-07-11  None  3.015715e-04
101  CGINS-CTDMOG-11646       CC_a1 2017-11-21  None  3.003766e-04
162  CGINS-CTDMOG-12638       CC_a1 2015-06-08  None  3.064611e-04
163  CGINS-CTDMOG-12638       CC_a1 2016-09-21  None  2.994633e-04
29   CGINS-CTDMOG-10255       CC_a1 2017-09-06  None  3.066937e-04
27   CGINS-CTDMOG-10255       CC_a1 2012-12-11  None  3.162308e-04
220  CGINS-CTDMOG-11638       CC_a1 2016-07-04  None  3.070477e-04
219  CGINS-CTDMOG-11638       CC_a1 2017-09-01  None  3.012959e-04
282  CGINS-CTDMOG-13422       CC_a1 2015-08-18  None  3.024181e-04
283  CGINS-CTDMOG-13422       CC_a1 2016-10-15  None  3.113964e-04
221  CGINS-CTDMOG-11638       CC_a1 2014-06-20  None  2.998744e-04
99   CGINS-CTDMOG-11646       CC_a1 2014-06-20  None  3.086916e-04
215  CGINS-CTDMOG-11638       CC_a2 2014-06-20  None -3.979843e-06
214  CGINS-CTDMOG-11638       CC_a2 2016-07-04  None -4.541412e-06
213  CGINS-CTDMOG-11638       CC_a2 2017-09-01  None -4.074699e-06
159  CGINS-CTDMOG-12638       CC_a2 2015-06-08  None -4.333575e-06
..                  ...         ...        ...   ...           ...
34   CGINS-CTDMOG-10255  CC_ptempa1 2015-11-10  None  5.137354e-02
166  CGINS-CTDMOG-12638  CC_ptempa1 2016-09-21  None  5.269413e-02
167  CGINS-CTDMOG-12638  CC_ptempa1 2015-06-08  None  5.208854e-02
35   CGINS-CTDMOG-10255  CC_ptempa1 2017-09-06  None  5.156059e-02
31   CGINS-CTDMOG-10255  CC_ptempa2 2012-12-11  None -5.100431e-07
30   CGINS-CTDMOG-10255  CC_ptempa2 2015-11-10  None -4.461589e-07
104  CGINS-CTDMOG-11646  CC_ptempa2 2017-11-21  None -7.424525e-07
223  CGINS-CTDMOG-11638  CC_ptempa2 2017-09-01  None -5.049620e-07
103  CGINS-CTDMOG-11646  CC_ptempa2 2014-06-20  None -4.024401e-07
224  CGINS-CTDMOG-11638  CC_ptempa2 2014-06-20  None -4.462353e-07
222  CGINS-CTDMOG-11638  CC_ptempa2 2016-07-04  None -7.288322e-07
102  CGINS-CTDMOG-11646  CC_ptempa2 2016-07-11  None -4.635052e-07
284  CGINS-CTDMOG-13422  CC_ptempa2 2015-08-18  None -1.159737e-06
285  CGINS-CTDMOG-13422  CC_ptempa2 2016-10-15  None  2.248582e-06
165  CGINS-CTDMOG-12638  CC_ptempa2 2015-06-08  None -5.782849e-07
32   CGINS-CTDMOG-10255  CC_ptempa2 2017-09-06  None -5.111036e-07
164  CGINS-CTDMOG-12638  CC_ptempa2 2016-09-21  None -7.671162e-07
140  CGINS-CTDMOG-11646    CC_wbotc 2017-11-21  None -6.782600e-08
139  CGINS-CTDMOG-11646    CC_wbotc 2014-06-20  None -6.782600e-08
138  CGINS-CTDMOG-11646    CC_wbotc 2016-07-11  None -6.782576e-08
187  CGINS-CTDMOG-12638    CC_wbotc 2015-06-08  None  2.283800e-07
186  CGINS-CTDMOG-12638    CC_wbotc 2016-09-21  None  2.283800e-07
260  CGINS-CTDMOG-11638    CC_wbotc 2014-06-20  None -1.732000e-07
259  CGINS-CTDMOG-11638    CC_wbotc 2017-09-01  None -1.732000e-07
258  CGINS-CTDMOG-11638    CC_wbotc 2016-07-04  None -1.731956e-07
306  CGINS-CTDMOG-13422    CC_wbotc 2015-08-18  None  3.081500e-08
307  CGINS-CTDMOG-13422    CC_wbotc 2016-10-15  None  3.081500e-08
67   CGINS-CTDMOG-10255    CC_wbotc 2017-09-06  None  4.253800

Unfortunately, it turns out that most of these endpoints are not ideal, because they return all of the available calibrations for an asset, and not just those used for the deployment we're interested in.  In the example above, there are only 5 deployments, but we found over a dozen calibration sets.

The system uses the calibration values that are closest in time but before the deployment start date.  So really, for any given deployment, all we need to do is find the calibrations for that asset that occurred just prior to the deployment (though, this could be 6 months or more earlier).

Note also, that calibrations do not have an end date.  It is assumed they are valid until the next calibration for the specific asset. 

As an alternative, we can use the following request to retrieve the calibration values that were used by the system to process data on any given day (which should theoretically be the same across a given deployment).  If you keep the date range within one deployment, you should only get one set of calibration values.

In [44]:
# Setup the API request url
request_url = 'https://ooinet.oceanobservatories.org/api/m2m/12587/asset/cal'

params = {
  'beginDT':'2016-10-01T00:00:00.000Z',
  'endDT':'2016-11-01T00:00:00.000Z',
  'refdes':'GP03FLMB-RIM01-02-CTDMOG060'
}

# Grab the information from the server
r = requests.get(request_url, params=params, auth=(USERNAME, TOKEN))
data = r.json()

# Reformat the data into a pretty table
df = pd.DataFrame() # Setup empty array
for d in data[0]['sensor']['calibration']:
  for dd in d['calData']:
    df = df.append({
      'value': dd['value'],
      'start': convert_time(dd['eventStartTime']),
      'stop': convert_time(dd['eventStopTime']),
      'name': dd['eventName'],
      'assetUid': dd['assetUid'],
      }, ignore_index=True)
    
df = df.sort_values(by=['start','name'])
df

assetUid        name      start  stop         value
0   CGINS-CTDMOG-11638       CC_a0 2016-07-04  None -1.041285e-04
9   CGINS-CTDMOG-11638       CC_a1 2016-07-04  None  3.070477e-04
7   CGINS-CTDMOG-11638       CC_a2 2016-07-04  None -4.541412e-06
6   CGINS-CTDMOG-11638       CC_a3 2016-07-04  None  2.028125e-07
13  CGINS-CTDMOG-11638    CC_cpcor 2016-07-04  None -9.570000e-08
19  CGINS-CTDMOG-11638    CC_ctcor 2016-07-04  None  3.250000e-06
2   CGINS-CTDMOG-11638        CC_g 2016-07-04  None -9.797545e-01
4   CGINS-CTDMOG-11638        CC_h 2016-07-04  None  1.699942e-01
5   CGINS-CTDMOG-11638        CC_i 2016-07-04  None -4.481542e-04
18  CGINS-CTDMOG-11638        CC_j 2016-07-04  None  6.360808e-05
17  CGINS-CTDMOG-11638  CC_p_range 2016-07-04  None  1.450000e+03
14  CGINS-CTDMOG-11638      CC_pa0 2016-07-04  None  4.652784e-02
15  CGINS-CTDMOG-11638      CC_pa1 2016-07-04  None  4.448754e-03
16  CGINS-CTDMOG-11638      CC_pa2 2016-07-04  None -1.061555e-11
21  CGINS-CTDMOG-11638    CC_ptca0 2016-07-04  None  5.245439e+05
20  CGINS-CTDMOG-11638    CC_ptca1 2016-07-04  None -1.082292e+00
23  CGINS-CTDMOG-11638    CC_ptca2 2016-07-04  None  4.952659e-02
8   CGINS-CTDMOG-11638    CC_ptcb0 2016-07-04  None  2.530163e+01
1   CGINS-CTDMOG-11638    CC_ptcb1 2016-07-04  None -1.275000e-03
3   CGINS-CTDMOG-11638    CC_ptcb2 2016-07-04  None  0.000000e+00
12  CGINS-CTDMOG-11638  CC_ptempa0 2016-07-04  None -6.899079e+01
11  CGINS-CTDMOG-11638  CC_ptempa1 2016-07-04  None  5.231744e-02
10  CGINS-CTDMOG-11638  CC_ptempa2 2016-07-04  None -7.288322e-07
22  CGINS-CTDMOG-11638    CC_wbotc 2016-07-04  None -1.731956e-07

# Instrument Vocabulary
We can use the Vocabulary API endpoint to convert a given reference designator for an instrument into its descriptive names.

In [45]:
# Retrieve vocabulary information for a given instrument
VOCAB_BASE_URL = 'https://ooinet.oceanobservatories.org/api/m2m/12586/vocab/inv'
request_url = VOCAB_BASE_URL+'/GP03FLMB/RIM01/02-CTDMOG060'
r = requests.get(request_url, auth=(USERNAME, TOKEN))
data = r.json()

data

[{'@class': '.VocabRecord',
  'model': 'SBE 37IM',
  'manufacturer': 'Sea-Bird',
  'vocabId': 1098,
  'refdes': 'GP03FLMB-RIM01-02-CTDMOG060',
  'instrument': 'CTD (30 meters)',
  'tocL1': 'Global Station Papa',
  'tocL2': 'Flanking Subsurface Mooring B',
  'tocL3': 'Mooring Riser',
  'mindepth': 30.0,
  'maxdepth': 30.0}]

# Annotations
Finally, let's grab all of the annotations for this instrument.  The annotations endpoint requires all of the options to be specified in the params array.

In [46]:
pd.set_option('display.max_colwidth', 30) # Show the full annotation text

In [47]:
ANNO_API = 'https://ooinet.oceanobservatories.org/api/m2m/12580/anno/find'
params = {
  'beginDT': int(datetime.date(2010,1,1).strftime('%s'))*1000,
  'endDT': int(datetime.date(2018,1,1).strftime('%s'))*1000,
  'refdes': 'GP03FLMB-RIM01-02-CTDMOG060',
#   'method': 'telemetered',
#   'stream': 'ctdmo_ghqr_sio_mule_instrument',
}

r = requests.get(ANNO_API, params=params, auth=(USERNAME, TOKEN))
data = r.json()

df = pd.DataFrame() # Setup empty array
for d in data:
  df = df.append({
    'annotation': d['annotation'],
    'start': convert_time(d['beginDT']),
    'stop': convert_time(d['endDT']),
    'site': d['subsite'],
    'node': d['node'],
    'sensor': d['sensor'],
    'id': d['id']
  }, ignore_index=True)

pd.set_option('display.max_colwidth', -1) # Show the full annotation text
df = df.sort_values(by=['start'])
df

annotation  \
1  Deployment 1: Mooring was deployed 5-10m shallower than planned.                                                                        
4  Deployment 2: Mooring was deployed 10-15m shallower than planned. Upon recovery of platform, biofouling was apparent on shallow CTDs.   
5  Deployment 2: Recovered data are not available for this deployment.                                                                     
2  Deployment 3: Mooring was deployed 10m deeper than planned.                                                                             
3  Deployment 4: Mooring was deployed 10m deeper than planned.                                                                             
0  Deployment 5: No data will be available via satellite telemetry. Data will be available when the mooring is recovered.                  

      id   node        sensor      site               start  \
1  154.0  None   None          GP03FLMB 2013-07-24 06:32:00   
4  155.0  None   None          GP03FLMB 2014-06-20 02:00:00   
5  584.0  RIM01  02-CTDMOG060  GP03FLMB 2014-06-20 02:00:00   
2  156.0  None   None          GP03FLMB 2015-06-08 21:25:00   
3  157.0  None   None          GP03FLMB 2016-07-04 23:07:00   
0  142.0  None   None          GP03FLMB 2017-07-14 22:49:00   

                  stop  
1  2014-06-18 00:00:00  
4  2015-06-07 00:00:00  
5  2015-06-07 00:00:00  
2  2016-07-03 18:35:00  
3  2017-07-17 15:02:00  
0  None

# Resources
For more information on using the API, you can check out:

* [OOI Machine 2 Machine Introduction](http://oceanobservatories.org/ooi-m2m-interface/)
* [OOI Machine to Machine (M2M) Basics](https://github.com/ooi-data-review/m2m_demo/blob/master/notebooks/basic_examples.ipynb)
* [Accessing data as JSON via OOI RESTful API](https://github.com/ooi-data-review/m2m_demo/blob/master/notebooks/json_data_request.ipynb)
* [Accessing data as NetCDF via OOI RESTful API](https://github.com/ooi-data-review/m2m_demo/blob/master/notebooks/netcdf_data_request.ipynb)
* [OOI Machine to Machine (M2M) Realtime Requests](https://github.com/ooi-data-review/m2m_demo/blob/master/notebooks/realtime_requests.ipynb)
